proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+```text
+parseWithoutDest (parse into GC heap, stay in GC heap ***): OK
+  358  ms ± 6.8 ms, 196 MB allocated, 402 MB copied, 441 MB peak memory
+
+parseWithoutDest (parse into GC heap, then copy into compact region): OK
+  332  ms ± 4.4 ms, 327 MB allocated, 362 MB copied, 377 MB peak memory
+
+parseWithDest (parse directly into compact region): OK
+  136  ms ± 5.4 ms, 283 MB allocated,  37 MB copied, 269 MB peak memory
+```
+
+(`***`_The  program only parses the document and then shutdowns, so in a real-world use case doing something non trivial with the parsed data, the time spent by the GC traversing the parsed dataset again and again in `parseWithoutDest (parse into GC heap, stay in GC heap` would be significantly more important._)
+
+Unlike the previous example, here the code for the parser is very similar and equally verbose for both the naive and DPS version, so there isn't much extra effort needed to get these performance benefits.
+
+## Effect and Interactions
+
+Primitive `compactAddHollow#` allow for direct allocation in a compact region, without copy, as stated in the [Motivation](#motivation) section. It needs to be combined with `reifyInfoTablePtr#` to properly allocate requested constructor objects with the right info table pointer, because as a RTS primitive, `compactAddHollow#` doesn't have access itself to the info table pointer to set for the constructor (all information about info tables has long been inlined/erased when the primitive is actually executed).
+
+`LiftedCtorToSymbol` is used to ease the development of a safe interface around `compactAddHollow#`/`reifyInfoTablePtr#`. In particular, [it allows to filter out the `Generic` representation of a type to find information about fields a specified data constructor](#get-info-about-fields-of-a-data-constructor-using-ghcgenerics-and-liftedctortosymbol) when the constructor is only known by its type-level, lifted representation (which is the one used by `compactAddHollow#`/`reifyInfoTablePtr#`).
+
+This set of primops is flexible enough to allow safe destination-passing style programming in Haskell, which allows for implementation of all examples presented above.
+
+I don't think the proposed changes will have any specific interaction with other GHC features. The implementation of the new primitives makes very little assumptions on the existing internals of GHC.
+
+`compactAddHollow#` is more unsafe than existing `compactAdd#` and `compactAddWithSharing#` primitives, but as are many primitives from `GHC.Prim`. That being said, the compact region guarantees wrt. the garbage collector should be enough to ensure a defined behavior for any program using this new primitive.
```

Comment:
Not if you explicitly read from unallocated memory, though. You just know that the GC isn't going to mess your game up.